feat: make torch-ttnn pip installable with optional ttnn dependency

aybanda opened the pull request:
## Ticket

Fixes #1036

## Problem Description

Previously, using the `pytorch2.0_ttnn` package required users to clone the repository and manually add it to their `PYTHONPATH`. This PR fixes the packaging workflow to allow users to easily install the package using `pip install torch-ttnn`.

## What’s Changed

- **Packaging:**  
  Updated `setup.py` and `pyproject.toml` to enable building a standard Python wheel and source distribution, allowing for installation via pip.
- **Dependencies:**  
  - Reverted all dependencies to match the current main branch for maximum compatibility.
  - The hardware-specific `ttnn` library is now an optional dependency.
  - Users can install the basic package with `pip install torch-ttnn`.
  - For supported Linux systems with Python 3.10, the `ttnn` backend can be installed via `pip install torch-ttnn[ttnn]`.

- **Documentation:**  
  Updated `README.md` with clear installation instructions reflecting these changes.
- **Workflows:**  
  Aligned CI workflows and dependency markers with upstream main and maintainer’s latest commits.

## Outcome

These changes fulfill the requirement of making `torch-ttnn` pip installable, significantly improving the developer experience and ensuring consistency with the main branch and upstream best practices.

jmalone-tt commented on setup.py:
```diff
@@ -1,22 +1,153 @@
+    ],
+    extras_require={
+        "ttnn": [
+            "ttnn>=0.1.0; platform_system=='Linux' and python_version=='3.10'",
```
For `ttnn` we currently rely on an updated URL for release candidates. This is not ideal, but keeps us in sync with their project. The URL should be updated in `requirements-dev.txt` so if we can use that, the problem will be fixed. Otherwise, we may need to update the GitHub Actions workflow that updates dependencies.

jmalone-tt reviewed (changes requested):
I verified that running `python -m build` correctly builds the wheel again - thanks for your work on this!

I left a couple of questions and clarifications.

Please also update the workflows that automatically build and push a wheel file to PyPI (`.github/workflows/update-ttnn-wheel.yaml` and `.github/actions/common_wheel_install/action.yaml` for example)

jmalone-tt commented on README.md:
```diff
@@ -5,6 +5,27 @@
+
+### Requirements
+- Python 3.10
+- Linux operating system
```
We also currently only support Ubuntu officially - would be good to update throughout the PR

aybanda commented:
Thanks for the review! @jmalone-tt  I've removed CUDA requirement, specified Ubuntu support, centralized dependencies in requirements.txt, and moved TTNN URL to requirements-dev.txt while maintaining package functionality.

aybanda commented:
@jmalone-tt  the validate-pr check was successful before the branch update - could you please review the latest changes and approve the workflow run if everything looks good?

jmalone-tt reviewed (changes requested):
Thanks for the updates! I just went through and left some more notes about tt-metal version updates. Once that is fixed, I will run through all the workflows and make sure everything is working as expected!

aybanda commented:
@jmalone-tt I've synced the TTNN version across requirements.txt, setup.py, and pyproject.toml and added a workflow to automate updates. This should resolve the version mismatch issues you pointed out and make future updates easier.

jmalone-tt commented:
Thanks for the updates here! I will kick off some of the workflows today and will send updates with the results

jmalone-tt commented:
Looks like the URL for tt-metal is incorrect. I think you should be able to trigger workflow runs and someone else on the team or I can approve them for debugging going forward:
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/15697463349/job/44225252051

dgomezTT commented on MANIFEST.in:
```diff
@@ -0,0 +1,18 @@
+include VERSION
+
+# Exclude test files from the distribution
+prune tests
```
just curious, can't you just use a pattern? like test/* or something like that?

dgomezTT reviewed (approved):
looks great to me!

jmalone-tt commented:
Ok, I believe it is set up with the trusted publisher system, so it should just work now. Please try again and let me know if there are any problems

jmalone-tt commented on .github/workflows/update-ttnn-wheel.yaml:
```diff
@@ -66,14 +79,14 @@ jobs:
         run: gh pr merge --merge --auto "${{ steps.create-pr.outputs.pull-request-number }}"
         env:
-          GH_TOKEN: ${{ secrets.GH_TOKEN }}
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Same comment as above

jmalone-tt commented:
Thanks for your patience, I was OOO at the end of last week. I just kicked off another run here:
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/15827143832

aybanda commented:
Hey @jmalone-tt  — I believe the commit that was approved and tested may not have included the latest fix. I’ve since reverted the `.whl` filename in `requirements.txt` back to the correct variant, which resolves the 404 errors.

jmalone-tt commented:
Just ran with latest commits. Looks like there are some issues with the Build, Test, and Release Wheel workflow:
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/15831101758

aybanda commented:
Addressed most of the issues 
Can you give it another go

jmalone-tt commented:
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/15833833895
Still at least an issue with the test-wheel-smoke step

jmalone-tt commented:
There are a minimal set of runners available, so I kicked off a run here:
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/15862439761

The main change is adding the line to checkout the repo at the beginning of the smoke tests. It looks like there might be an issue with the latest tt-metal URL. Let's try just using the one that's in main today and I can update tt-metal version in a separate PR

aybanda commented:
Thanks for kicking off the rerun and isolating the issue. Sticking with the `tt-metal` version from `main` sounds good. From the logs, the telemetry-related errors seem consistent with earlier runs. Let me know if you'd like me to tweak anything further.

jmalone-tt commented:
Looks like there is an issue with the test-wheel-model step. You should be able to refer to the model-tests step in run-tests to get this working. Please also integrate the changes to add `+cpu` to the `torch` and `torchvision` lines of requirements.txt and revert to the tt-metal version currently in main

jmalone-tt commented on requirements-dev.txt:
```diff
@@ -1,4 +1,3 @@
--r requirements.txt
```
This will break many of our CI workflows. Do we really need to remove this line for PyPI compatibility? If so, it will require updating workflows too, but I would prefer to just revert this change

aybanda commented:
check this out @jmalone-tt 
https://test.pypi.org/project/torch-ttnn-test/0.0.1.dev710/

aybanda commented:
removed the direct URL dependency as PyPI does not allow packages with URL-based requirements.

aybanda commented:
> Ok, it sounds like we will need to reconfigure how some of the release process / dependency install works for the project. I can handle that in another PR, so you don't need to worry about it.
> 
> For now, let's just get the release passing successfully pointing to one of the tt-metal releases on PyPI. For this PR, let's point to https://pypi.org/project/ttnn/ version 0.59.0 (0.60.0 doesn't work with this project yet)
> 
> Additional details: requirements.txt should still have the URL for ttnn (and it should still be updated by the update-ttnn-wheel workflow) You can keep the URL removed from the pyproject.toml and remove references to requirements.txt or requirements-dev.txt from the pyproject.toml Ultimately, the hope is that people can install torch-ttnn[ttnn] and get a working version of both, with the ttnn version defined by the latest official release on PyPI
> 
> Please let me know if there are any more questions and I will be happy to help!

Updated requirements.txt to include the platform-specific ttnn wheel URL for CI, while keeping setup.py and pyproject.toml PyPI-compliant with only ttnn==0.59.0 as an optional extra.

jmalone-tt commented:
Thanks!
Run started here:
https://github.com/tenstorrent/pytorch2.0_ttnn/actions/runs/16060016555

jmalone-tt commented on requirements.txt:
```diff
@@ -6,5 +6,4 @@ networkx==3.1
 paramiko==3.5.1
-
-ttnn @ https://github.com/tenstorrent/tt-metal/releases/download/v0.59.0-rc47/ttnn-0.59.0rc47-cp310-cp310-manylinux_2_34_x86_64.whl ; python_version=="3.10"
+ttnn @ https://github.com/tenstorrent/tt-metal/releases/download/v0.59.0-rc31/ttnn-0.59.0rc31-cp310-cp310-linux_x86_64.whl ; python_version=="3.10"
```
Please change back to 0.59.0rc47

aybanda commented:
You're welcome!
We're using setuptools_scm for versioning — it's configured in pyproject.toml with:

```
[tool.setuptools_scm]  
version_scheme = "python-simplified-semver"  
local_scheme = "no-local-version"  
write_to = "torch_ttnn/_version.py"
```

This auto-generates the version from Git tags. So, for PyPI uploads, we just need to push a tag like v0.0.1 or v0.1.0 — and that becomes the published version.
